_posts/2022-04-02-gcp-finops.md

```diff
@@ -0,0 +1,386 @@
+GCP SKUs are very detailed (which is generally fine), for example every type of CPU is reported under different SKU.
+But it introduces accidental complexity, on the presented dashboards everything is aggregated under single entry "CPU".
+More convenient to use, when you want to focus on the costs not on the technical details.
+
+### BigQuery, Pubsub and Cloud Storage reports
+
+Dataproc or Dataflow are only a part of total data pipeline costs.
+The next three dashboards are for the cost tracking of the most important "storages" used by the pipelines:
+
+* BigQuery used by batch and streaming pipelines
+* Pubsub used by streaming pipelines
+* Cloud storage used mainly by batch pipelines as a staging area
+
+On the BigQuery costs dashboard you can filter by the dataset name.
+It is very important when the data pipeline stores results into more than one dataset.
+The filtering is available only if the **allegro__dataset_name** label was set.
+
+![BigQuery dashboard](/assets/images/finops_bigquery.webp)
+
+On the Pubsub costs dashboard you can filter by the topic name.
+It is very important when the data pipeline publishes to or subscribes on many topics.
+The filtering is available only if the **allegro__topic_name** label was set.
+
+![Pubsub dashboard](/assets/images/finops_pubsub.webp)
+
+On the Cloud Storage dashboard you can filter by the bucket name.
+It is very important when the data pipeline stores results into more than one bucket.
+The filtering is available only if the **allegro__bucket_name** label was set.
+
+![Cloud Storage dashboard](/assets/images/finops_gcs.webp)
+
+## The challenges
+
+Right now you should have a much better understanding of the difference between "resources" and "data pipelines" oriented cost reports.
+The final reports may look relatively simple but this is easier said than done.
+
+* How to automate labeling to avoid gaps in the reports?
+* What if the GCP product does not support labels at all?
+* What if the labels set on the resources for unknown reasons are not available in the billing export?
+* How to track costs of resources created automatically by managed GCP services? 
+  Even if we set labels on the service itself they are not propagated to the underlying resources.
+* What if you are using the libraries like Apache Beam or Apache Spark, and the API for setting labels is not publicly exposed?
+* There is also a shared infrastructure like Cloud Composer or Cloud Logging, they also generate significant costs.
+
+Let's tackle all identified challenges one by one.
+
+### Labeling automation
+
+* Use [Terraform](https://www.terraform.io) for "static" cloud resources management like BigQuery datasets, Pubsub topics, Pubsub subscriptions and Cloud storage buckets.
+Prepare the reusable modules with input parameters required to set the mandatory labels.
+If Terraform is too low-level, create some [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) on top. 
```

Comment:
I think currently more appropriate suggestion is to use [Terraform CDK](https://github.com/hashicorp/terraform-cdk), as creating your own DSL is a bit of a hassle

Reply:
CDK is still very low-level, just another language instead of TF. DSL could be a REAL abstraction.